openedx/core/djangoapps/oauth_dispatch/docs/decisions/0003-use-jwt-as-oauth-tokens-remove-openid-connect.rst

```diff
@@ -0,0 +1,154 @@
+and external partner services. In addition, there are multiple edX microservices
+that are OAuth2 Clients of the LMS.
+
+Some of the internal microservice clients require `OpenID Connect`_ features.
+Specifically, they make use of the `ID Token`_ extension to get user profile
+details from the LMS via the OAuth protocol. The ID Token can also be forwarded
+from one microservice to another, allowing the recipient microservice to
+validate the identity of the token's owner without needing to reconnect with a
+centralized LMS.
+
+We have integrated our fork of DOP_ with support for OpenID Connect. So, an
+access_token request with a DOP client::
+
+    curl -X POST -d "client_id=abc&client_secret=def&grant_type=client_credentials" http://localhost:18000/oauth2/access_token/
+
+includes an id_token field::
+
+    {
+        "access_token": <RANDOMLY-GENERATED-ACCESS-TOKEN>,
+        "id_token": <BASE64-ENCODED-ID-TOKEN>,
+        "expires_in": 31535999,
+        "token_type": "Bearer",
+        "scope": "profile openid email permissions"
+    }
+
+where the value of BASE64-ENCODED-ID-TOKEN decodes to::
+
+    {
+        "family_name": "User1",
+        "administrator": false,
+        "sub": "foo",
+        "iss": "http://localhost:18000/oauth2",
+        "user_tracking_id": 1234,
+        "preferred_username": "user1",
+        "name": "User 1",
+        "locale": "en",
+        "given_name": "User 1",
+        "exp": 1516757075,
+        "iat": 1516753475,
+        "email": "user1@edx.org",
+        "aud": "bar"
+    }
+
+However, OpenID Connect is a large standard with many features and is not supported by
+the DOT_ implementation.
+
+.. _OpenID Connect: http://openid.net/connect/
+.. _ID Token: http://openid.net/specs/openid-connect-core-1_0.html#IDToken
+.. _DOP: https://github.com/caffeinehit/django-oauth2-provider
+.. _DOT: https://github.com/evonove/django-oauth-toolkit
+
+Decision
+--------
+
+Remove our dependency on OpenID Connect since we don't really need all its
+features and it isn't supported by DOT. Instead, support `JSON Web Token (JWT)`_,
+which is a simpler standard and integrates well with the OAuth2 protocol.
+
+.. _JSON Web Token (JWT): https://jwt.io/
+
+The simplest approach is to allow OAuth2 clients to request JWT tokens in place
+of randomly generated Bearer tokens. JWT tokens contain user information,
+replacing the need for OpenID's ID Tokens altogether.
+
+JWT Authentication Library
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Use the open source `Django Rest Framework JWT library`_ as the backend
+implementation for JWT token type authentication.
+
+.. _Django Rest Framework JWT library: https://getblimp.github.io/django-rest-framework-jwt/
+
+JWT Tokens
+~~~~~~~~~~
+
+An OAuth2 client requesting a JWT token_type::
+
+    curl -X POST -d "client_id=abc&client_secret=def&grant_type=client_credentials&token_type=jwt" http://localhost:18000/oauth2/access_token/
```

Comment:
Based on the RFC spec (https://tools.ietf.org/html/rfc7523) I believe token_type should still be `bearer` not `jwt` and grant_type would be `urn:ietf:params:oauth:grant-type:jwt-bearer`

Sticking with the spec should allow better library compatibility for both us and Open edX, and seems to match what others are doing:
- https://developer.atlassian.com/cloud/jira/platform/oauth-2-jwt-bearer-token-authorization-grant-type/
- https://help.salesforce.com/articleView?id=remoteaccess_oauth_jwt_flow.htm&type=5

Reply:
I had also discovered that our implementation is "unique" when doing the research for these docs.

However, in digging into this further, I found that the [RFC for OAuth JWT Assertions](https://tools.ietf.org/html/rfc7523#section-2.1) is for a different purpose entirely. It talks about using JWTs as a replacement for the [assertion](https://tools.ietf.org/html/rfc7521) in the OAuth handshake.  That is, it uses the JWT as a means to *get an OAuth token* (instead of using [traditional OAuth grant types](https://tools.ietf.org/html/rfc6749#section-4) that require *client-secrets* or *passwords*).  For us, however, we are using a JWT *in place of an OAuth token*.

If we supported the [IETF standard](https://tools.ietf.org/html/rfc7523#section-2.1), client Applications would never need to send their client-secrets over-the-wire when requesting an OAuth Token.  Instead, they would use the out-of-band-exchanged client-secret to sign their own JWT. The self-created JWT presents an *assertion* about the client Application. This would be a stronger story for authenticating client Application requests.  We could still support this in the future, if we wanted.

In our case here, however, the Authorization server (LMS) creates/signs a JWT to bind information about the requesting application and the authorizing user.  This self-contained token can then be validated/used by any relying party (microservice/API) for granting access.

Although I couldn't find an RFC or IETF standard for it, we are using a relatively maintained and used [open-source library](https://getblimp.github.io/django-rest-framework-jwt) for this implementation.

Thanks for raising this concern.  It is worth putting in the docs.